File: android/app/src/main/java/com/bor_devs/shoplist/data/sync/WriteQueue.kt
package com.bor_devs.shoplist.data.sync

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

enum class WriteKind { UPDATE, DELETE }

data class WriteOp(
    val collection: String,
    val recordId: String,
    val kind: WriteKind,
    val body: Map<String, Any?>? = null,
    val attempts: Int = 0,
)

/**
 * Generic per-record write queue with coalescing and retry/backoff, ported 1:1
 * from web/src/lib/writeQueue.ts. A later enqueue for the same record merges
 * with the pending op (a delete supersedes updates); transient failures retry
 * with backoff; a permanently rejected write invokes [onGaveUp].
 */
class WriteQueue(
    private val scope: CoroutineScope,
    private val retryDelaysMs: List<Long> = listOf(1000, 3000, 9000),
    private val isOnline: () -> Boolean,
    private val execute: suspend (WriteOp) -> Any?,
    private val onApplied: (WriteOp, Any?) -> Unit = { _, _ -> },
    private val onGaveUp: (WriteOp, Throwable) -> Unit = { _, _ -> },
    private val shouldRetry: (Throwable, WriteOp) -> Boolean = { _, _ -> true },
) {
    private val lock = Mutex()
    private val ops = LinkedHashMap<String, WriteOp>()
    private val timers = HashMap<String, Job>()
    private val running = HashSet<String>()

    private fun keyOf(collection: String, recordId: String) = "$collection:$recordId"

    fun enqueueUpdate(collection: String, recordId: String, body: Map<String, Any?>) {
        scope.launch {
            lock.withLock {
                val key = keyOf(collection, recordId)
                val existing = ops[key]
                if (existing?.kind == WriteKind.DELETE) return@withLock // delete supersedes
                ops[key] = WriteOp(
                    collection, recordId, WriteKind.UPDATE,
                    body = (existing?.body ?: emptyMap()) + body,
                    attempts = 0,
                )
                schedule(key, 0)
            }
        }
    }

    fun enqueueDelete(collection: String, recordId: String) {
        scope.launch {
            lock.withLock {
                val key = keyOf(collection, recordId)
                ops[key] = WriteOp(collection, recordId, WriteKind.DELETE, attempts = 0)
                schedule(key, 0)
            }
        }
    }

    fun flush() {
        scope.launch {
            lock.withLock {
                for ((key, op) in ops) {
                    ops[key] = op.copy(attempts = 0)
                    schedule(key, 0)
                }
            }
        }
    }

    suspend fun size(): Int = lock.withLock { ops.size }

    // must be called holding [lock]
    private fun schedule(key: String, delayMs: Long) {
        timers[key]?.cancel()
        timers[key] = scope.launch {
            if (delayMs > 0) delay(delayMs)
            lock.withLock { timers.remove(key) }
            run(key)
        }
    }

    private suspend fun run(key: String) {
        val op: WriteOp = lock.withLock {
            if (running.contains(key)) return
            val current = ops[key] ?: return
            if (!isOnline()) return // flush() re-schedules on reconnect
            running.add(key)
            ops.remove(key) // claim
            current
        }

        try {
            val result = execute(op)
            onApplied(op, result)
        } catch (error: Throwable) {
            val retriable = shouldRetry(error, op)
            lock.withLock {
                if (retriable && op.attempts < retryDelaysMs.size) {
                    val requeued = ops[key]
                    if (requeued != null) {
                        // Newer intent queued meanwhile; keep its fields on top.
                        if (op.kind == WriteKind.UPDATE && requeued.kind == WriteKind.UPDATE) {
                            ops[key] = requeued.copy(body = (op.body ?: emptyMap()) + (requeued.body ?: emptyMap()))
                        }
                    } else {
                        ops[key] = op.copy(attempts = op.attempts + 1)
                    }
                    schedule(key, retryDelaysMs[minOf(op.attempts, retryDelaysMs.size - 1)])
                } else if (!ops.containsKey(key)) {
                    onGaveUp(op, error)
                }
            }
        } finally {
            lock.withLock {
                running.remove(key)
                if (ops.containsKey(key) && !timers.containsKey(key)) schedule(key, 0)
            }
        }
    }
}

File: web/src/lib/writeQueue.ts
// Generic per-record write queue with coalescing and retry/backoff.
// Pure (no app imports) so it can be unit tested; the app wires a concrete
// instance in itemWriteQueue.ts.

export interface WriteOp {
    collection: string;
    recordId: string;
    kind: 'update' | 'delete';
    body?: Record<string, unknown>;
    attempts: number;
}

export interface WriteQueue {
    enqueueUpdate(collection: string, recordId: string, body: Record<string, unknown>): void;
    enqueueDelete(collection: string, recordId: string): void;
    flush(): void;
    size(): number;
    has(collection: string, recordId: string): boolean;
}

const DEFAULT_RETRY_DELAYS = [1000, 3000, 9000];

export function createWriteQueue(opts: {
    execute: (op: WriteOp) => Promise<unknown>;
    onApplied?: (op: WriteOp, result: unknown) => void;
    onGaveUp?: (op: WriteOp, error: unknown) => void;
    shouldRetry?: (error: unknown, op: WriteOp) => boolean;
    isOnline?: () => boolean;
    retryDelays?: number[];
}): WriteQueue {
    const retryDelays = opts.retryDelays ?? DEFAULT_RETRY_DELAYS;
    const isOnline = opts.isOnline ?? (() => typeof navigator === 'undefined' || navigator.onLine);

    const ops = new Map<string, WriteOp>();
    const timers = new Map<string, ReturnType<typeof setTimeout>>();
    const running = new Set<string>();

    const keyOf = (collection: string, recordId: string) => `${collection}:${recordId}`;

    function schedule(key: string, delay: number) {
        const existing = timers.get(key);
        if (existing) clearTimeout(existing);
        timers.set(key, setTimeout(() => {
            timers.delete(key);
            void run(key);
        }, delay));
    }

    async function run(key: string) {
        // An active run reschedules pending work in its finally block.
        if (running.has(key)) return;
        const op = ops.get(key);
        if (!op) return;
        if (!isOnline()) return; // flush() re-schedules on reconnect

        running.add(key);
        // Claim the op: anything enqueued while the request is in flight
        // lands in the map as fresh work instead of mutating this op.
        ops.delete(key);
        try {
            const result = await opts.execute(op);
            opts.onApplied?.(op, result);
        } catch (error) {
            const retriable = opts.shouldRetry ? opts.shouldRetry(error, op) : true;
            if (retriable && op.attempts < retryDelays.length) {
                const requeued = ops.get(key);
                if (requeued) {
                    // Newer intent was queued meanwhile; keep its fields on top.
                    if (op.kind === 'update' && requeued.kind === 'update') {
                        requeued.body = { ...op.body, ...requeued.body };
                    }
                } else {
                    ops.set(key, { ...op, attempts: op.attempts + 1 });
                }
                schedule(key, retryDelays[Math.min(op.attempts, retryDelays.length - 1)]);
            } else if (!ops.has(key)) {
                opts.onGaveUp?.(op, error);
            }
        } finally {
            running.delete(key);
            if (ops.has(key) && !timers.has(key)) schedule(key, 0);
        }
    }

    return {
        enqueueUpdate(collection, recordId, body) {
            const key = keyOf(collection, recordId);
            const existing = ops.get(key);
            if (existing?.kind === 'delete') return; // deletion supersedes updates
            ops.set(key, {
                collection,
                recordId,
                kind: 'update',
                body: { ...(existing?.body || {}), ...body },
                attempts: 0
            });
            schedule(key, 0);
        },
        enqueueDelete(collection, recordId) {
            const key = keyOf(collection, recordId);
            ops.set(key, { collection, recordId, kind: 'delete', attempts: 0 });
            schedule(key, 0);
        },
        flush() {
            for (const [key, op] of ops) {
                op.attempts = 0;
                schedule(key, 0);
            }
        },
        size: () => ops.size,
        has: (collection, recordId) => ops.has(keyOf(collection, recordId))
    };
}
